syntax = "proto3";
package mama.tile;

message DistanceTable {
    // sorted edge ids
    repeated uint32 edge_id = 1;
    // distance in meters from edge which owns this table to the edge with correponding edge_id
    repeated uint32 distance = 2;
}

message Edge {
    uint32 length = 1;

    // if < 0, then the edge leaves tile and (abs(...) - 1) correponds to the object in the `neighbour_tile_nodes`
    sint32 target_node_id = 2;

    // id of shape in the `shape_spatial_index`
    uint32 shape_id = 3;

    // optional distance table
    DistanceTable distance_table = 4;
}

message Node {
    repeated uint32 adjacent_edges = 2;
}

message Header {
    repeated Edge edges = 1;
    repeated Node nodes = 2;
    // contains (tile_id << 32) | node_id
    repeated uint64 neighbour_tile_node_ids = 3;
    bytes shape_spatial_index = 4;
}
